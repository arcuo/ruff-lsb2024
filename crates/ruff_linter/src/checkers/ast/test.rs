use std::path::Path;

use ruff_python_index::Indexer;
use ruff_python_parser::{lexer::LexResult, Mode};
use ruff_python_semantic::Module;
use ruff_source_file::Locator;

use crate::{
    directives, linter::TokenSource, packaging::detect_package_root, settings::{flags::Noqa, LinterSettings}
};

use super::Checker;

// TODO: Should we use this?

/// Create a new [`Checker`] for the given source code
pub(crate) fn test_checker_state_after_snippet_traversal(source: &str) -> &Checker {
    let path = Path::new("");
    let locator = Locator::new(source);
    let tokens = TokenSource::Tokens(ruff_python_parser::tokenize_all(source, Mode::Module));
    let indexer = Indexer::from_tokens(&tokens, &locator);
    let settings = LinterSettings::default();
    let directives = directives::extract_directives(
        &tokens,
        directives::Flags::from_settings(&settings),
        &locator,
        &indexer,
    );
    let checker = Checker::new(
        &settings,
        &directives.noqa_line_for,
        Noqa::Disabled,
        &path,
        path.parent()
            .and_then(|parent| detect_package_root(parent, &settings.namespace_packages)),
        Module {
            kind: ruff_python_semantic::ModuleKind::Module,
            source: ruff_python_semantic::ModuleSource::Path(&[path.to_str().unwrap().to_string()]),
            python_ast: tokens.,
            name: path.to_string(),
        },
    );
    &checker
}
