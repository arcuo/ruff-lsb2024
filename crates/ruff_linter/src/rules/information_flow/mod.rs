//! Rules for information flow.
pub(crate) mod rules;

#[cfg(test)]
mod tests {
    use std::path::Path;

    use anyhow::Result;
    use test_case::test_case;

    use crate::registry::Rule;
    use crate::test::test_path;
    use crate::{assert_messages, settings};

    // Basic rules
    #[test_case(Rule::IFMustIncludeVariableLabel, Path::new("IF001.py"))]
    #[test_case(Rule::IFMissingPrincipal, Path::new("IF002.py"))]
    // Explicit
    #[test_case(Rule::IFExplicitVariableAssign, Path::new("IF101.py"))] // TODO: split into multiple tests
    #[test_case(Rule::IFExplicitVariableAssign, Path::new("IF101_for.py"))] // TODO: split into multiple tests

    // Implicit
    #[test_case(Rule::IFImplicitVariableAssign, Path::new("IF201_if.py"))]
    #[test_case(Rule::IFImplicitVariableAssign, Path::new("IF201_while.py"))]
    #[test_case(Rule::IFImplicitVariableAssign, Path::new("IF201_for.py"))]
    #[test_case(Rule::IFImplicitVariableAssign, Path::new("IF201_nested.py"))]
    #[test_case(Rule::IFImplicitVariableAssign, Path::new("IF201_assert.py"))]

    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.noqa_code(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("information_flow").join(path).as_path(),
            &settings::LinterSettings::for_rule(rule_code),
        )?;
        assert_messages!(snapshot, diagnostics);
        Ok(())
    }
}
