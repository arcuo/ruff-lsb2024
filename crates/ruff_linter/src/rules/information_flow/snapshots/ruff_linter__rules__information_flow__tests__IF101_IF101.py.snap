---
source: crates/ruff_linter/src/rules/information_flow/mod.rs
---
IF101.py:18:1: IF101 Inconfidential assignment to more restrictive variable. Expression `bob_var` with label `{ bob }` is being assigned to `bob_var = alice_bob_var
` with label `{ alice, bob }`
   |
17 | # IF101: Fail - Information flow from {alice,bob} to {bob}
18 | bob_var = alice_bob_var
   | ^^^^^^^ IF101
19 | 
20 | # IF101: Fail - public var are less restrictive than alice_bob_var
   |

IF101.py:21:1: IF101 Inconfidential assignment to more restrictive variable. Expression `public_var` with label `{}` is being assigned to `public_var = alice_bob_var
` with label `{ alice, bob }`
   |
20 | # IF101: Fail - public var are less restrictive than alice_bob_var
21 | public_var = alice_bob_var
   | ^^^^^^^^^^ IF101
22 | 
23 | # IF101: Fail - public var are less restrictive than bob_var
   |

IF101.py:24:1: IF101 Inconfidential assignment to more restrictive variable. Expression `public_var` with label `{}` is being assigned to `public_var = bob_var
` with label `{ bob }`
   |
23 | # IF101: Fail - public var are less restrictive than bob_var
24 | public_var = bob_var
   | ^^^^^^^^^^ IF101
25 | 
26 | # IF101: Success - Public variables can flow to more restrictive variables
   |

IF101.py:37:1: IF101 Inconfidential assignment to more restrictive variable. Expression `bob_var` with label `{ bob }` is being assigned to `bob_var = alice_bob_var + 1
` with label `{ alice, bob }`
   |
36 | # # IF101: Fail Information flow from b to a
37 | bob_var = alice_bob_var + 1
   | ^^^^^^^ IF101
38 | 
39 | # # IF101: Success - Information flow from a to b
   |

IF101.py:52:1: IF101 Inconfidential assignment to more restrictive variable. Expression `bob_var` with label `{ bob }` is being assigned to `bob_var += alice_bob_var
` with label `{ alice, bob }`
   |
51 | # IF101: Fail Information flow from b to a
52 | bob_var += alice_bob_var
   | ^^^^^^^ IF101
53 | 
54 | # IF101: Fail Information flow from b to a
   |

IF101.py:55:1: IF101 Inconfidential assignment to more restrictive variable. Expression `public_var` with label `{}` is being assigned to `public_var += bob_var
` with label `{ bob }`
   |
54 | # IF101: Fail Information flow from b to a
55 | public_var += bob_var
   | ^^^^^^^^^^ IF101
56 | 
57 | ## Annotated assignment - i.e ": int ="
   |

IF101.py:63:1: IF101 Inconfidential assignment to more restrictive variable. Expression `bob_var` with label `{ bob }` is being assigned to `bob_var: int = alice_bob_var
` with label `{ alice, bob }`
   |
62 | # IF101: Fail Information flow from b to a
63 | bob_var: int = alice_bob_var
   | ^^^^^^^ IF101
64 | 
65 | # IF101: Success - Information flow from a to b
   |

IF101.py:74:1: IF101 Inconfidential assignment to more restrictive variable. Expression `public_var` with label `{}` is being assigned to `public_var = bob_var = alice_bob_var
` with label `{ alice, bob }`
   |
73 | # IF101: Fail - Information flow from {} to {bob}, {alice,bob}
74 | public_var = bob_var = alice_bob_var
   | ^^^^^^^^^^ IF101
75 | 
76 | ## Tuple assignment
   |

IF101.py:74:14: IF101 Inconfidential assignment to more restrictive variable. Expression `bob_var` with label `{ bob }` is being assigned to `public_var = bob_var = alice_bob_var
` with label `{ alice, bob }`
   |
73 | # IF101: Fail - Information flow from {} to {bob}, {alice,bob}
74 | public_var = bob_var = alice_bob_var
   |              ^^^^^^^ IF101
75 | 
76 | ## Tuple assignment
   |

IF101.py:84:16: IF101 Inconfidential assignment to more restrictive variable. Expression `bob_var` with label `{ bob }` is being assigned to `alice_bob_var, bob_var = 1, alice_bob_var
` with label `{ alice, bob }`
   |
83 | # IF101: Fail - Information flow from {alice,bob} to {bob}
84 | alice_bob_var, bob_var = 1, alice_bob_var
   |                ^^^^^^^ IF101
85 | alice_bob_var, bob_var = alice_bob_var
   |

IF101.py:85:16: IF101 Inconfidential assignment to more restrictive variable. Expression `bob_var` with label `{ bob }` is being assigned to `alice_bob_var, bob_var = alice_bob_var
` with label `{ alice, bob }`
   |
83 | # IF101: Fail - Information flow from {alice,bob} to {bob}
84 | alice_bob_var, bob_var = 1, alice_bob_var
85 | alice_bob_var, bob_var = alice_bob_var
   |                ^^^^^^^ IF101
   |

IF101.py:94:1: IF101 Inconfidential assignment to more restrictive variable. Expression `bob_var` with label `{ bob }` is being assigned to `bob_var = [1, 2, alice_bob_var]
` with label `{ alice, bob }`
   |
93 | # IF101: Fail - Information flow from {alice,bob} to {bob}
94 | bob_var = [1, 2, alice_bob_var]
   | ^^^^^^^ IF101
95 | 
96 | # IF101: Fail - Information flow from {bob} to {public}
   |

IF101.py:97:1: IF101 Inconfidential assignment to more restrictive variable. Expression `public_var` with label `{}` is being assigned to `public_var = [1, 2, bob_var]
` with label `{ bob }`
   |
96 | # IF101: Fail - Information flow from {bob} to {public}
97 | public_var = [1, 2, bob_var]
   | ^^^^^^^^^^ IF101
98 | 
99 | # Tuple assignment
   |

IF101.py:105:1: IF101 Inconfidential assignment to more restrictive variable. Expression `bob_var` with label `{ bob }` is being assigned to `bob_var = (1, 2, alice_bob_var)
` with label `{ alice, bob }`
    |
104 | # IF101: Fail - Information flow from {alice,bob} to {bob}
105 | bob_var = (1, 2, alice_bob_var)
    | ^^^^^^^ IF101
106 | 
107 | # IF101: Fail - Information flow from {bob} to {public}
    |

IF101.py:108:1: IF101 Inconfidential assignment to more restrictive variable. Expression `public_var` with label `{}` is being assigned to `public_var = (1, 2, bob_var)
` with label `{ bob }`
    |
107 | # IF101: Fail - Information flow from {bob} to {public}
108 | public_var = (1, 2, bob_var)
    | ^^^^^^^^^^ IF101
109 | 
110 | # Set assignment
    |

IF101.py:116:1: IF101 Inconfidential assignment to more restrictive variable. Expression `bob_var` with label `{ bob }` is being assigned to `bob_var = {1, 2, alice_bob_var}
` with label `{ alice, bob }`
    |
115 | # IF101: Fail - Information flow from {alice,bob} to {bob}
116 | bob_var = {1, 2, alice_bob_var}
    | ^^^^^^^ IF101
117 | 
118 | # IF101: Fail - Information flow from {bob} to {public}
    |

IF101.py:119:1: IF101 Inconfidential assignment to more restrictive variable. Expression `public_var` with label `{}` is being assigned to `public_var = {1, 2, bob_var}
` with label `{ bob }`
    |
118 | # IF101: Fail - Information flow from {bob} to {public}
119 | public_var = {1, 2, bob_var}
    | ^^^^^^^^^^ IF101
120 | 
121 | # Dict assignment
    |

IF101.py:127:1: IF101 Inconfidential assignment to more restrictive variable. Expression `bob_var` with label `{ bob }` is being assigned to `bob_var = {1: 2, 3: alice_bob_var}
` with label `{ alice, bob }`
    |
126 | # IF101: Fail - Information flow from {alice,bob} to {bob}
127 | bob_var = {1: 2, 3: alice_bob_var}
    | ^^^^^^^ IF101
128 | 
129 | # IF101: Fail - Information flow from {bob} to {public}
    |

IF101.py:130:1: IF101 Inconfidential assignment to more restrictive variable. Expression `public_var` with label `{}` is being assigned to `public_var = {1: 2, 3: bob_var}
` with label `{ bob }`
    |
129 | # IF101: Fail - Information flow from {bob} to {public}
130 | public_var = {1: 2, 3: bob_var}
    | ^^^^^^^^^^ IF101
131 | 
132 | ## TODO: var = func() Might be a separate rule
    |
